In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [54]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_market_data)

In [55]:
# Create a DataFrame with the scaled data
df_scaled = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns
)

# Copy the crypto names from the original DataFrame
df_scaled["coin_id"] = df_market_data.index

# Set the coin_id column as index
df_scaled = df_scaled.set_index("coin_id")

# Display the scaled DataFrame
df_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [56]:
# Create a list with the number of k-values from 1 to 11
k_original = list(range(1, 11))

In [57]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k_original:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    model_original = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_scaled`
    model_original.fit(df_scaled)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(model_original.inertia_)


c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:14

In [ ]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_original = {
    "k": k_original,
    "inertia": inertia
}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow_original = pd.DataFrame(elbow_data_original)

In [15]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k_values
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [16]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [17]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_scaled)

c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4)

In [18]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = model.predict(df_scaled)

# Print the resulting array of cluster values.
print(clusters)

[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [19]:
# Create a copy of the scaled DataFrame
df_scaled_predictions = df_scaled.copy()

In [20]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled_predictions["Clusters"] = clusters

# Display the copy of the scaled DataFrame
df_scaled_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Clusters  
coin_id                                             
bitcoin                        -0.251637         3  
ethereum                       -0.199352         3  
tether                         -0.282061         0  
ripple                         -0.295546         0  
bitcoin-cash                   -0.270317         3

In [23]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
df_scaled_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="Clusters",
    hover_cols=["coin_id"]
)

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Clusters,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [24]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [29]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_data = pca.fit_transform(df_scaled)

# View the scaled PCA data
pca_data

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292],
       [-0.24819846, -1.37625159, -1.46202571],
       [-0.43840762, -0.17533654, -0.6633884 ],
       [-0.69342533, -0.47381462, -0.52759693],
       [ 0.06049915,  2.90940385,  1.49857131],
       [-0.39335243, -0.10819197, -0.01275608],
       [-0.79617564, -0.49440875,  1.08281169],
       [ 0.06407452, -1.26982514, -1.09882928],
       [-0.48901506, -0.73271912, -0.06254323],
       [-0.3062723 ,  0.70341515,  1.71422359],
       [-0.51352775, -0.14280239, -0.65656583],
       [-0.36212044, -0.98691441, -0.728

In [30]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_

print("Explained Variance Ratio:")
for i, variance in enumerate(explained_variance, start=1):
    print(f"PC{i}: {variance:.4f}")

Explained Variance Ratio:
PC1: 0.3720
PC2: 0.3470
PC3: 0.1760


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** Total Explained Variance = 0.3720 + 0.3470 + 0.1760 = 0.8950 = 89.5%

In [31]:
# Create a new DataFrame with the PCA data.
df_pca = pd.DataFrame(
    pca_data,
    columns=["PC1", "PC2", "PC3"]
)

# Copy the crypto names from the original scaled DataFrame
df_pca["coin_id"] = df_scaled.index

# Set the coin_id column as index
df_pca = df_pca.set_index("coin_id")

# Display the scaled PCA DataFrame
df_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [33]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [34]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_pca`
    model.fit(df_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)

c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:14

In [35]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k": k,
    "inertia": inertia
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)


In [36]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve (PCA)", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Still 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Although the inertia values are slightly different, the overall shape of the graph is generally the same, with the ideal value for k still being 4.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [37]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [38]:
# Fit the K-Means model using the PCA data
model.fit(df_pca)

c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4)

In [39]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_clusters = model.predict(df_pca)

# Print the resulting array of cluster values.
print(pca_clusters)

[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [40]:
# Create a copy of the scaled PCA DataFrame
df_pca_clustered = df_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_pca_clustered["Clusters"] = pca_clusters

# Display the copy of the scaled PCA DataFrame
df_pca_clustered.head()

PC1       PC2       PC3  Clusters
coin_id                                             
bitcoin      -0.600667  0.842760  0.461595         3
ethereum     -0.458261  0.458466  0.952877         3
tether       -0.433070 -0.168126 -0.641752         0
ripple       -0.471835 -0.222660 -0.479053         0
bitcoin-cash -1.157800  2.041209  1.859715         3

In [41]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
df_pca_clustered.hvplot.scatter(
    x="PC1",
    y="PC2",
    c="Clusters",
    hover_cols=["coin_id"]
)

:Scatter   [PC1]   (PC2,Clusters,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [ ]:
#pip install jupyter_bokeh

Note: you may need to restart the kernel to use updated packages.


In [47]:
df_elbow["k"] = df_elbow["k"].astype(int)
df_elbow_pca["k"] = df_elbow_pca["k"].astype(int)

In [49]:
# Composite plot to contrast the Elbow curves
df_elbow["Source"] = "Original"
df_elbow_pca["Source"] = "PCA Reduced"
df_elbow_combined = pd.concat([df_elbow, df_elbow_pca])
df_elbow_combined.hvplot.line(
    x="k",
    y="inertia",
    by="Source",
    title="Elbow Curve Comparison: Original vs PCA-Reduced Data",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    xticks=list(range(1, 11))
)

:NdOverlay   [Source]
   :Curve   [k]   (inertia)

In [52]:
# Composite plot to contrast the clusters
df_original_clusters = df_scaled.copy()
df_original_clusters["Cluster"] = model.predict(df_scaled)
df_original_clusters["Source"] = "Original"
df_original_clusters["PC1"] = df_pca["PC1"]
df_original_clusters["PC2"] = df_pca["PC2"]

# PCA clustering results
df_pca_clustered = df_pca.copy()
df_pca_clustered["Cluster"] = model.predict(df_pca)
df_pca_clustered["Source"] = "PCA Reduced"

# Combine both DataFrames for comparison
df_cluster_comparison = pd.concat([df_original_clusters, df_pca_clustered], ignore_index=True)


df_cluster_comparison.hvplot.scatter(
    x="PC1",
    y="PC2",
    groupby="Source",
    c="Cluster",
    hover_cols=["coin_id"],
    title="Comparison of Clusters: Original vs PCA-Reduced Data"
)

ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- price_change_percentage_14d
- price_change_percentage_1y
- price_change_percentage_200d
- price_change_percentage_24h
- price_change_percentage_30d
- ...
Feature names seen at fit time, yet now missing:
- PC1
- PC2
- PC3


In [53]:
# Create and fit a new KMeans model on PCA-reduced data
pca_kmeans = KMeans(n_clusters=4, random_state=1)
pca_kmeans.fit(df_pca)

# Predict clusters based on PCA data
df_pca_clustered = df_pca.copy()
df_pca_clustered["Cluster"] = pca_kmeans.predict(df_pca)
df_pca_clustered["Source"] = "PCA Reduced"

df_original_clusters = df_scaled.copy()
df_original_clusters["Cluster"] = model.predict(df_scaled)
df_original_clusters["Source"] = "Original"
df_original_clusters["PC1"] = df_pca["PC1"]
df_original_clusters["PC2"] = df_pca["PC2"]

df_cluster_comparison = pd.concat([df_original_clusters, df_pca_clustered], ignore_index=True)

# Plot composite cluster comparison
df_cluster_comparison.hvplot.scatter(
    x="PC1",
    y="PC2",
    groupby="Source",
    c="Cluster",
    cmap="Category10",
    hover_cols=["coin_id"],
    title="Cluster Comparison: Original vs PCA-Reduced"
)

c:\Users\jenn2\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- price_change_percentage_14d
- price_change_percentage_1y
- price_change_percentage_200d
- price_change_percentage_24h
- price_change_percentage_30d
- ...
Feature names seen at fit time, yet now missing:
- PC1
- PC2
- PC3


#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 